chia/pools/pool_puzzles.py

```diff
@@ -342,20 +352,38 @@ def solution_to_extra_data(full_spend: CoinSolution) -> Optional[PoolState]:
 
     if full_spend.coin.puzzle_hash == SINGLETON_LAUNCHER_HASH:
         # Launcher spend
-        extra_data = full_solution.rest().rest().first().first().as_atom()
+        extra_data: Program = full_solution.rest().rest().first().first().as_atom()
         return PoolState.from_bytes(extra_data)
 
     # Not launcher spend
     inner_solution: Program = full_solution.rest().rest().first()
-    inner_spend_type: int = inner_solution.first().as_int()
-
-    if inner_spend_type == 0:
-        # Absorb
-        return None
 
     # Spend which is not absorb, and is not the launcher
-    extra_data = inner_solution.rest().rest().rest().first().as_atom()
-    return PoolState.from_bytes(extra_data)
+    num_args = len(inner_solution.as_python())
+    assert num_args in (2, 3)
+
+    if num_args == 2:
+        # pool member
+        if inner_solution.rest().first().as_int() == 1:
+            return None
+        extra_data = inner_solution.first()
+        state_bytes: Optional[bytes] = None
+        for key, value in extra_data.as_python():
```

Comment:
Please factor this code out, since it will be used in many places (for example, it's used twice in just this function!).